File: 78-subsets/subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        Stack<Integer> s = new Stack<>();
        dfs(nums, 0, s, result);

        return result;
    }

    public void dfs(int[] nums, int i, Stack<Integer> s, List<List<Integer>> result) {
        if (i == nums.length) {
            result.add(new ArrayList(s));
            return;
        }
        s.push(nums[i]);
        dfs(nums, i + 1, s, result);
        s.pop();
        dfs(nums, i + 1, s, result);
    }
}

File: 78-subsets/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def helper(nums: List[int], i: int, s: List[int]):
            if i == len(nums):
                result.add(tuple(s))
                return
            t = s[:]
            t.append(nums[i])
            helper(nums, i + 1, t)
            t.remove(nums[i])
            helper(nums, i + 1, t)

        result = set()
        helper(nums, 0, [])
        return result